src/mon/ConnectionTracker.cc

```diff
@@ -0,0 +1,256 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+/*
+ * Ceph - scalable distributed file system
+ *
+ * Copyright (C) 2019 Red Hat
+ *
+ * This is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License version 2.1, as published by the Free Software
+ * Foundation.  See file COPYING.
+ *
+ */
+
+#include "ConnectionTracker.h"
+#include "common/Formatter.h"
+
+std::ostream& operator<<(std::ostream&o, const ConnectionReport& c) {
+  o << "rank=" << c.rank << ",epoch=" << c.epoch << ",version=" << c.epoch_version
+    << ", current links: " << c.current << ", history: " << c.history;
+  return o;
+}
+
+std::ostream& operator<<(std::ostream& o, const ConnectionTracker& c) {
+  o << "rank=" << c.rank << ", epoch=" << c.epoch << ", version=" << c.version
+    << ", half_life=" << c.half_life << ", reports: " << c.peer_reports;
+  return o;
+}
+
+ConnectionReport *ConnectionTracker::reports(int p)
+{
+  auto i = peer_reports.find(p);
+  if (i == peer_reports.end()) {
+    ceph_assert(p != rank);
+    auto[j,k] = peer_reports.insert(std::pair<int,ConnectionReport>(p,ConnectionReport()));
+    i = j;
+  }
+  return &i->second;
+}
+
+const ConnectionReport *ConnectionTracker::reports(int p) const
+{
+  auto i = peer_reports.find(p);
+  if (i == peer_reports.end()) {
+    return NULL;
+  }
+  return &i->second;
+}
+
+void ConnectionTracker::receive_peer_report(const ConnectionReport& report)
+{
+  if (report.rank == rank) return;
+  encoding.clear();
```

Comment:
It's moved since you asked to compress the functions (so it's only called once per ConnectionTracker peer report), but I'm not going to try and make this conditional — I just prefer keeping things simple when possible and the cost is not very high.